### PR TITLE
Merge the newlibc string into NuttX.

### PR DESCRIPTION
**Pull request by yangguangcai01**
## Summary
Merge the newlibc string into NuttX.This is jjohnstn@redhat.com response regarding copyright:
Hello,

Individual files in newlib either have a license in them or they fall under the first license found in the COPYING.NEWLIB file which is in the top level directory:

(1) Red Hat Incorporated

Copyright (c) 1994-2009  Red Hat, Inc. All rights reserved.

This copyrighted material is made available to anyone wishing to use,
modify, copy, or redistribute it subject to the terms and conditions
of the BSD License.   This program is distributed in the hope that
it will be useful, but WITHOUT ANY WARRANTY expressed or implied,
including the implied warranties of MERCHANTABILITY or FITNESS FOR
A PARTICULAR PURPOSE.  A copy of this license is available at
http://www.opensource.org/licenses. Any Red Hat trademarks that are
incorporated in the source code or documentation are not subject to
the BSD License and may only be used or replicated with the express
permission of Red Hat, Inc.

-- Jeff J.
## Impact
use optimizate c string.
## Testing
ci

### CODE REVIEW

**Review by acassis — COMMENTED** [review]
@yangguangcai01 did you check the license? newlibc has BSD and LGPL code mixed. Also even if you are adding only BSD licensed code you need to identify it

**Comment by xiaoxiang781216** [comment]
All code come from https://github.com/bminor/newlib/tree/master/newlib/libc/string which doesn't declare copyright in source code. should we copy the full text from to https://github.com/bminor/newlib/blob/master/COPYING.NEWLIB to https://github.com/apache/nuttx/blob/master/LICENSE?

**Comment by xiaoxiang781216** [comment]
@yangguangcai01 let add copyright to LICENSE file

**Comment by raiden00pl** [comment]
if this code has a different licence than NuttX then the build of these files must be guarded by `CONFIG_ALLOW_XXX_COMPONENTS`. Adding copyrights to the LICENSE file is not sufficient. But in case of this PR where files have no copyright headers, what option should be used ?

**Comment by jerpelea** [comment]
If there is no license in the file they inherit the license from LICENSE file. 
I would advise to handle with care files that go into the core NuttX since they may cause legal issues. 

All this code should be scanned with a snippet scanner to ensure that we import the licenses for code that we think that we import and avoid pollution.

**Review by jerpelea — COMMENTED** [review]
I would like to see a snippet scanner + FOSSID scan result to make sure that we are importing what we think that we are importing 

If there is an alternative code implementing the same functionality I think that we should avoid the current code

File: libs/libc/string/lib_memrchr.c [hunk 1]
```diff
@@ -26,6 +26,34 @@
 
 #include <string.h>
 
+/****************************************************************************
```
Comment (jerpelea) [inline libs/libc/string/lib_memrchr.c:29]: code with a different license than the one in the header should not be imported in the same file.

**Review by jerpelea — CHANGES_REQUESTED** [review]

File: libs/libc/string/lib_memrchr.c (hunk 1 above)
Comment (GUIDINGLI) [reply libs/libc/string/lib_memrchr.c:31]: Got the newlib author response, see the Summary

**Review by raiden00pl — CHANGES_REQUESTED** [review]
but what is the final license of these files? is it BSD or Apache? Is there anyone who can clearly confirm this?

if it's BSD, it should depend on the  `ALLOW_BSD_COMPONENTS` option, but since these are basic OS features, the entire OS should depend on the BSD by default

**Comment by GUIDINGLI** [comment]
> but what is the final license of these files? is it BSD or Apache? Is there anyone who can clearly confirm this?
> 
> if it's BSD, it should depend on the `ALLOW_BSD_COMPONENTS` option, but since these are basic OS features, the entire OS should depend on the BSD by default

As the Summary said, "BSD License"
We will update this PR, add 'ALLOW_BSD_COMPONENTS' depends.

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@raiden00pl @jerpelea All license concern is addressed, please review again.

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by raiden00pl** [comment]
what do you thing about creating a new configuration option like this:

```
config LIBC_STRING_OPTIMIZE
  depends on ALLOW_BSD_COMPONENTS
  default y
  --help--
    Use optimized string function implementation based on newlib.
```

this way, it'll be more visible to the user that feature like this exist. On the other hand, the user may want to disable this optimization (I assume it is more FLASH consuming).

**Comment by yangguangcai01** [comment]
> what do you thing about creating a new configuration option like this:
> 
> ```
> config LIBC_STRING_OPTIMIZE
>   depends on ALLOW_BSD_COMPONENTS
>   default y
>   --help--
>     Use optimized string function implementation based on newlib.
> ```
> 
> this way, it'll be more visible to the user that feature like this exist. On the other hand, the user may want to disable this optimization (I assume it is more FLASH consuming).

ok, done

**Review by raiden00pl — APPROVED** [review]

**Comment by cederom** [comment]
Houston we gotta problem, I did a bisect from master head and looks like we broke configure with this one :-) 

Kconfig file syntax seems invalid and CI did not catch it :-)

@yangguangcai01 can you please double check?

@xiaoxiang781216 @raiden00pl shall we revert or patch ?

```
% git bisect good
6e87f110833363d353b7c3e849d510d72749b67d is the first bad commit
commit 6e87f110833363d353b7c3e849d510d72749b67d
Author: yangguangcai <yangguangcai@xiaomi.com>
Date:   Fri Jan 26 10:50:25 2024 +0800

    Merge the newlibc string into NuttX.

    Signed-off-by: yangguangcai <yangguangcai@xiaomi.com>

 LICENSE                          |  31 +++++++++++++++++++++++++++++++
 libs/libc/string/Kconfig         |   7 +++++++
 libs/libc/string/lib_memccpy.c   |  96 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_memchr.c    | 103 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_memcmp.c    |  69 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_memcpy.c    |  70 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_memmove.c   |   8 +-------
 libs/libc/string/lib_memrchr.c   | 103 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_stpcpy.c    |  48 ++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_stpncpy.c   |  72 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_strcat.c    |  53 +++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_strchr.c    | 113 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_strchrnul.c |   6 ++++++
 libs/libc/string/lib_strcmp.c    |  66 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_strcoll.c   |   2 +-
 libs/libc/string/lib_strcpy.c    |  53 +++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_strdup.c    |   2 +-
 libs/libc/string/lib_strlen.c    |  60 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_strncmp.c   |  85 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_strncpy.c   |  71 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 libs/libc/string/lib_strrchr.c   |  19 +++++++++++++++++++
 21 files changed, 1128 insertions(+), 9 deletions(-)


% git checkout master
% git checkout c113a224e8432cca8456b0f509b7716091594ee9

% git log -2
commit c113a224e8432cca8456b0f509b7716091594ee9 (HEAD)
Author: yangguangcai <yangguangcai@xiaomi.com>
Date:   Mon Feb 26 22:59:52 2024 +0800

    newlibc:skip asan check.

    Signed-off-by: yangguangcai <yangguangcai@xiaomi.com>

commit 6e87f110833363d353b7c3e849d510d72749b67d
Author: yangguangcai <yangguangcai@xiaomi.com>
Date:   Fri Jan 26 10:50:25 2024 +0800

    Merge the newlibc string into NuttX.

    Signed-off-by: yangguangcai <yangguangcai@xiaomi.com>


% ./tools/configure.sh -B nucleo-l432kc:nsh
  Copy files
  Select CONFIG_HOST_BSD=y
  Refreshing...
CP: arch/dummy/Kconfig to /XXX/nuttx.git/arch/dummy/dummy_kconfig
CP: boards/dummy/Kconfig to /XXX/nuttx.git/boards/dummy/dummy_kconfig
LN: platform/board to /XXX/nuttx-apps.git/platform/dummy
LN: include/arch to arch/arm/include
LN: include/arch/board to /XXX/nuttx.git/boards/arm/stm32l4/nucleo-l432kc/include
LN: drivers/platform to /XXX/nuttx.git/drivers/dummy
LN: include/arch/chip to /XXX/nuttx.git/arch/arm/include/stm32l4
LN: arch/arm/src/chip to /XXX/nuttx.git/arch/arm/src/stm32l4
LN: arch/arm/src/board to /XXX/nuttx.git/boards/arm/stm32l4/nucleo-l432kc/src
mkkconfig in /XXX/nuttx-apps.git/audioutils
mkkconfig in /XXX/nuttx-apps.git/benchmarks
mkkconfig in /XXX/nuttx-apps.git/boot
mkkconfig in /XXX/nuttx-apps.git/canutils
mkkconfig in /XXX/nuttx-apps.git/crypto
mkkconfig in /XXX/nuttx-apps.git/database
mkkconfig in /XXX/nuttx-apps.git/examples/mcuboot
mkkconfig in /XXX/nuttx-apps.git/examples/module
mkkconfig in /XXX/nuttx-apps.git/examples/sotest
mkkconfig in /XXX/nuttx-apps.git/examples
mkkconfig in /XXX/nuttx-apps.git/fsutils
mkkconfig in /XXX/nuttx-apps.git/games
mkkconfig in /XXX/nuttx-apps.git/graphics
mkkconfig in /XXX/nuttx-apps.git/industry
mkkconfig in /XXX/nuttx-apps.git/inertial
mkkconfig in /XXX/nuttx-apps.git/interpreters/luamodules
mkkconfig in /XXX/nuttx-apps.git/interpreters
mkkconfig in /XXX/nuttx-apps.git/logging
mkkconfig in /XXX/nuttx-apps.git/lte
mkkconfig in /XXX/nuttx-apps.git/math
mkkconfig in /XXX/nuttx-apps.git/mlearning
mkkconfig in /XXX/nuttx-apps.git/netutils
mkkconfig in /XXX/nuttx-apps.git/sdr
mkkconfig in /XXX/nuttx-apps.git/system
mkkconfig in /XXX/nuttx-apps.git/testing
mkkconfig in /XXX/nuttx-apps.git/videoutils
mkkconfig in /XXX/nuttx-apps.git/wireless/bluetooth
mkkconfig in /XXX/nuttx-apps.git/wireless/ieee802154
mkkconfig in /XXX/nuttx-apps.git/wireless
mkkconfig in /XXX/nuttx-apps.git
libs/libc/string/Kconfig:44: syntax error
libs/libc/string/Kconfig:43: unknown option "--help--"
libs/libc/string/Kconfig:44: unknown option "Use"
gmake: *** [tools/Unix.mk:696: olddefconfig] Error 1
ERROR: failed to refresh
```

**Comment by lupyuen** [comment]
Same here, all the RISC-V Builds are broken: https://gist.github.com/lupyuen/974f625cdd57da30805277566dd404e5

```text
libs/libc/string/Kconfig:44: syntax error
libs/libc/string/Kconfig:43: unknown option "--help--"
libs/libc/string/Kconfig:44: unknown option "Use"
make: *** [olddefconfig] Error 1
ERROR: failed to refresh
```

**Comment by lupyuen** [comment]
@yangguangcai01 We are missing hyphens in `---help---`:

```text
	--help--
		Use optimized string function implementation based on newlib.
```

Should be:

```text
	---help---
		Use optimized string function implementation based on newlib.
```

**Comment by cederom** [comment]
Thanks @lupyuen I have free moment here goes the fix https://github.com/apache/nuttx/pull/14388 :-)

**Comment by xiaoxiang781216** [comment]
@cederom do you find the reason why nuttx ci doesn't catch this type of error?

**Comment by lupyuen** [comment]
Wonder if CI is running a different version of Kconfig?

**Comment by yangguangcai01** [comment]
> @yangguangcai01 We are missing hyphens in `---help---`:
> 
> ```
> 	--help--
> 		Use optimized string function implementation based on newlib.
> ```
> 
> Should be:
> 
> ```
> 	---help---
> 		Use optimized string function implementation based on newlib.
> ```

Thanks for your thoughts.

**Comment by cederom** [comment]
I use `kconfig-frontends-4.11.0.1` on FreeBSD :-)

**Comment by xiaoxiang781216** [comment]
> Wonder if CI is running a different version of Kconfig?

ci may use https://pypi.org/project/kconfiglib/, we can switch some back to use kconfig-frontend.

**Comment by cederom** [comment]
Looking at the logs here: https://github.com/apache/nuttx/actions/runs/11384169288/job/31671394520#step:7:101

This script is used for builds: https://github.com/apache/nuttx/blob/master/tools/testbuild.sh

Comparing logs with the script it seems CMake is used for configuration not Kconfig ? #14285 https://github.com/apache/nuttx/blob/e75ac114c121b2194aeabe4f1a5bdefdafd847b4/tools/testbuild.sh#L507

**Comment by lupyuen** [comment]
I was experimenting with defconfig files: https://github.com/apache/nuttx/pull/14299

If the `tools/ci/testlist/arm-01.dat` file contains "CMake,nucleo-f334r8:adc", then CI Build runs CMake to apply the defconfig. Otherwise it calls `tools/configure.sh`

**Comment by xiaoxiang781216** [comment]
auto select from:
https://github.com/apache/nuttx/blob/master/tools/Unix.mk#L634-L679

**Comment by cederom** [comment]
> @xiaoxiang781216: auto select from: https://github.com/apache/nuttx/blob/master/tools/Unix.mk#L634-L679

Wow so kconfiglib can now work with all of the NuttX Kconfig files? :-)

**Comment by cederom** [comment]
0430AM here, setting sleep bit for a while, take care folks :-)

File: libs/libc/string/lib_memccpy.c [hunk 2]
```diff
@@ -51,6 +81,71 @@
...
+    {
+      if ((*pout++ = *pin++) == endchar)
+        {
+          ptr = pout;
+          break;
+        }
+    }
+
+  return ptr;
+#else
```
Comment (yamt) [inline libs/libc/string/lib_memccpy.c:148]: i feel we should separate code with different license to separate files. how do you think?

Comment (raiden00pl) [reply libs/libc/string/lib_memccpy.c:148]: +1

**Comment by acassis** [comment]
> > Wonder if CI is running a different version of Kconfig?
> 
> ci may use https://pypi.org/project/kconfiglib/, we can switch some back to use kconfig-frontend.

Good idea! kconfiglib doesn't have all features we have on kconfig-frontend, even the search doesn't work when you search for CONFIG_SYMBOL_NAME

File: libs/libc/string/lib_memccpy.c (hunk 2 above)
Comment (yangguangcai01) [reply libs/libc/string/lib_memccpy.c:148]: > i feel we should separate code with different license to separate files. how do you think?

ok, I will separate these files.

**Pull request merged**